### PR TITLE
feat(common): create a custom allocator template that statically allocates memory

### PR DESCRIPTION
**Pull request by Laura-Danielle**
## Overview
We did it!!

Alt Text

This is a first pass at creating a custom allocator template. It seems to generally work with some caveats (i.e. weird stuff happens with vector and deque). Ideally, we'll be able to use this allocator to help reduce repetitive code! For reference, we followed the [Allocator interface](https://en.cppreference.com/w/cpp/named_req/Allocator) to build out this class.

## Review Requests
- Take a look at naming, see if there are other tests (not listed in the to-do list) that might be good to have etc.

## To-Do

- [x] Figure out the exact requirements we aren't meeting for vector and deque (something is missing with the constructor type)
- [x] Add more detailed comments for everything
- [x] Add a test where the element is a class

### CODE REVIEW

File: include/common/core/allocator.hpp
Comment (fsinapi) [inline include/common/core/allocator.hpp:63]: Since there's no implementation here for `max_size()`, isn't there a chance that `allocate()` will be invoked to allocate more than one item at a time?

File: include/common/core/allocator.hpp
Comment (sfoster1) [inline include/common/core/allocator.hpp:44]: you sure about this ctor set? You've got the copy ctor defaulted, which means you can happily copy gigantic pools of memory around - which seems wrong - and then the copy-assign deleted, which is a weird combo.

File: include/common/core/allocator.hpp
Comment (sfoster1) [inline include/common/core/allocator.hpp:53]: You also need to handle unaligned deletes here with something like
```
auto wrapped = reinterpret_cast<AllocatorElement<Element>*>(loc);
if (wrapped < backing.begin()) return;
if ((wrapped - backing.begin()) % sizeof(*wrapped) != 0) return;
```
this shouldn't ever happen, but if something called `deallocate` with a wrongly-aligned address, then you wouldn't necessarily set an arbitrary byte, but you'd set the previous element which might or might not be right.

File: include/common/core/allocator.hpp
Comment (sfoster1) [inline include/common/core/allocator.hpp:80]: leftover comment?

File: include/common/core/allocator.hpp
Comment (sfoster1) [inline include/common/core/allocator.hpp:63]: not handling the `size_t` makes this not conformant and also means it can't handle array-allocations. This is probably why it doesn't work with `vector`, since that commonly works by proactively allocating array backings and reallocating larger ones when capacity is exceeded. IMO it's reasonable to scan the backing array for the shortest slot that's >= request.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Couple leftover comments and I think a mistakenly- ommitted `include/common/core/task.hpp`?

Plus I think not handling array-allocations is why `vector` doesn't work.

File: include/common/core/allocator.hpp
Comment (Laura-Danielle) [reply include/common/core/allocator.hpp:12]: Very old comment. Going to revamp the comments before merging.

File: include/common/core/allocator.hpp
Comment (Laura-Danielle) [reply include/common/core/allocator.hpp:63]: Hm I don't think allocate would be attempting to allocate more than one item, but we should definitely be checking that the size given doesn't exceed the max size of the element. Will probably end up doing something like [seth suggested](https://github.com/Opentrons/ot3-firmware/pull/376#discussion_r886798397).

Comment (fsinapi) [reply include/common/core/allocator.hpp:63]: `vector`, `deque`, and some other container types allocate arrays instead of individual items, which means `allocate` would attempt to allocate more than one item.

I think that the type of your `backing` array will have to change if you want to support block allocation with Seth's suggestion. If you allocate out a block of multiple contiguous elements in the current backing array, you're giving the caller free rein to write over the `AllocatorElement::used` members.

Comment (sfoster1) [reply include/common/core/allocator.hpp:63]: Oh that's a very good point, you're right.

**Review by sfoster1 — APPROVED** [review]
Approved once you remove the empty `include/common/core/task.hpp`

File: include/common/core/allocator.hpp
Comment (fsinapi) [inline include/common/core/allocator.hpp:8]: Might want to fix up this comment, looks like the formatter chewed it up pretty bad

**Review by fsinapi — APPROVED** [review]
Changes look good!

**Pull request merged**
